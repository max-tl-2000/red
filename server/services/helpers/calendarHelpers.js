/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { toMoment, parseAsInTimezone, DATE_ISO_FORMAT, stringRepresentsADateWithoutTime } from '../../../common/helpers/moment-utils';
import loggerModule from '../../../common/helpers/logger';
import { getPropertyById } from '../properties';
import { formatSimpleAddress } from '../../../common/helpers/addressUtils';

const logger = loggerModule.child({ subType: 'calendarHelpers' });

export const NUMBER_OF_DAYS_SYNC = 60;
export const NUMBER_OF_PAST_DAYS_SYNC = 30;
export const TIMEZONE_SYNC = 'Etc/UTC';

const momentsHaveSameTime = ({ startMoment, endMoment, timezone }) => {
  const start = toMoment(startMoment, { timezone });
  const end = toMoment(endMoment, { timezone });

  return end.format('HH:mm:ss') === start.format('HH:mm:ss');
};

export const convertToMomentFromDateOnlyOrISOFormat = (date, { timezone }) => {
  if (stringRepresentsADateWithoutTime(date)) {
    return parseAsInTimezone(date, { timezone });
  }

  return toMoment(date, { timezone });
};

// format received from Cronofy for full-day event:
// {
//   "calendar_id": "cal_U9uuErStTG@EAAAB_IsAsykA2DBTWqQTf-f0kJw",
//   "start": "2019-01-28",
//   "end": "2014-01-29",
//   "free_busy_status": "busy",
//   "event_uid": "evt_external_5c40e8fdd646c74f367684d5",
// },
//
// format received from Cronofy for an event which is not full-day:
// {
//   "calendar_id": "cal_U9uuErStTG@EAAAB_IsAsykA2DBTWqQTf-f0kJw",
//   "start": "2019-01-28T08:00:00Z",
//   "end": "2019-01-28T08:30:00Z",
//   "free_busy_status": "busy",
//   "event_uid": "evt_external_5c40e8fdd646c74f367684d5",
// },
export const isAllDayEvent = (event, timezone) => {
  const start = event.start || event.startDate;
  const end = event.end || event.endDate;

  if (start === end) return false;

  const startMoment = convertToMomentFromDateOnlyOrISOFormat(start, { timezone });
  const endMoment = convertToMomentFromDateOnlyOrISOFormat(end, { timezone });

  if (!momentsHaveSameTime({ startMoment, endMoment, timezone })) return false;

  return true;
};

export const adjustDatesToTimezoneForCronofyEvent = (ctx, event, timezone) => {
  const { start, end } = event;

  const isAllDay = isAllDayEvent(event, timezone);
  if (!isAllDay) return event;

  logger.trace({ ctx, cronofyEvent: event, timezone }, 'adjust event dates to timezone - input');
  const startDateTime = parseAsInTimezone(start, { timezone, format: DATE_ISO_FORMAT });
  const endDateTime = parseAsInTimezone(end, { timezone, format: DATE_ISO_FORMAT });
  const fixedEvent = {
    ...event,
    start: startDateTime.toISOString(),
    end: endDateTime.toISOString(),
  };
  logger.trace({ ctx, cronofyEvent: event, fixedEvent }, 'adjust event dates to timezone - output');
  return fixedEvent;
};

export const getDateRangeForSync = (startDate, numberOfDays) => {
  const dateMoment = toMoment(startDate, { parseFormat: DATE_ISO_FORMAT, timezone: TIMEZONE_SYNC, strict: false });
  const from = dateMoment.format();
  const to = dateMoment.add(numberOfDays, 'day').format();

  return { from, to };
};

export const getAppointmentAddress = async (ctx, appointment) => {
  logger.trace({ ctx, appointment }, 'getAppointmentAddress - params');

  if (!appointment.metadata || !appointment.metadata.selectedPropertyId) return '';

  const property = await getPropertyById(ctx, appointment.metadata.selectedPropertyId);

  return property.leasingOfficeAddress || formatSimpleAddress(property.address);
};
