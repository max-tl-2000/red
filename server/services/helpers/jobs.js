/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { getRecurringJobByName, getJobsInProgressByName, getPendingJobsByName } from '../jobs';
import { DALTypes } from '../../../common/enums/DALTypes';

export const jobIsInProgress = async (ctx, name) => {
  const { status: recurringJobStatus } = await getRecurringJobByName(ctx, name);
  const jobsInProgress = await getJobsInProgressByName(ctx, name);
  const pendingJobs = await getPendingJobsByName(ctx, name);
  return recurringJobStatus === DALTypes.JobStatus.IN_PROGRESS || jobsInProgress.length || pendingJobs.length;
};
