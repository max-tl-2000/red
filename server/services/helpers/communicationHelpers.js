/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { adjustWalkinDates } from '../../../common/helpers/walkInUtils';
import { DALTypes } from '../../../common/enums/DALTypes';
import { toMoment } from '../../../common/helpers/moment-utils';
import tryParse from '../../../common/helpers/try-parse';

export const getFirstComm = comms =>
  adjustWalkinDates(comms)
    .filter(comm => comm.direction === DALTypes.CommunicationDirection.OUT && comm.category === DALTypes.CommunicationCategory.USER_COMMUNICATION)
    .sort((a, b) => toMoment(a.created_at).diff(toMoment(b.created_at)))[0];

export const narrowDownPartiesByProperty = ({ parties, propertyId }) => {
  const filteredPartiesByProperty = parties.filter(p => p.assignedPropertyId === propertyId);
  return filteredPartiesByProperty.length ? filteredPartiesByProperty : parties;
};

export const extractPostMessage = post => {
  if (post.category === DALTypes.PostCategory.EMERGENCY) return post.message;

  /*
Sample raw post message from rich text editor
{
  "blocks": [
      { "key": "o385", "text": "normal text", "type": "unstyled", "depth": 0, "inlineStyleRanges": [], "entityRanges": [], "data": {} },
      { "key": "c8n6s", "text": "", "type": "unstyled", "depth": 0, "inlineStyleRanges": [], "entityRanges": [], "data": {} },
      { "key": "3nug8", "text": "bold text", "type": "unstyled", "depth": 0, "inlineStyleRanges": [{ "offset": 0, "length": 9, "style": "BOLD" }], "entityRanges": [], "data": {} },
      { "key": "27l3j", "text": "", "type": "unstyled", "depth": 0, "inlineStyleRanges": [], "entityRanges": [], "data": {} },
      { "key": "9al05", "text": "italic text", "type": "unstyled", "depth": 0, "inlineStyleRanges": [{ "offset": 0, "length": 6, "style": "ITALIC" }], "entityRanges": [], "data": {} },
      { "key": "brj3m", "text": "", "type": "unstyled", "depth": 0, "inlineStyleRanges": [], "entityRanges": [], "data": {} },
      { "key": "cvomr", "text": "link", "type": "unstyled", "depth": 0, "inlineStyleRanges": [], "entityRanges": [{ "offset": 0, "length": 4, "key": 0 }], "data": {} }
  ],
  "entityMap": { "0": { "type": "LINK", "mutability": "MUTABLE", "data": { "url": "https://rxp1.local.env.reva.tech/cohortComms" } } }
}
 */

  const { blocks } = tryParse(post.rawMessage, { blocks: [] });
  return blocks.map(b => b.text).join('\n');
};
