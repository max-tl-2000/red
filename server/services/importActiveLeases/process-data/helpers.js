/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import isEqual from 'lodash/isEqual';
import { DALTypes } from '../../../../common/enums/DALTypes';
import trim from '../../../../common/helpers/trim';
import { now, toMoment, parseAsInTimezone } from '../../../../common/helpers/moment-utils';
import { YEAR_MONTH_DAY_FORMAT } from '../../../../common/date-constants';
import loggerModule from '../../../../common/helpers/logger';
import { archiveParty } from '../../party';
import { validatePhone as validatePhoneNumber, getOnlyDigitsFromPhoneNumber } from '../../../../common/helpers/phone-utils';
import config from '../../../config';
import { enhanceContactInfoWithSmsInfo } from '../../telephony/twilio';
import { removeSpaces } from '../../../dal/helpers/person';
import { getActiveLeaseWorkflowDataByPartyId } from '../../../dal/activeLeaseWorkflowRepo';
import { archiveExternalInfoByPartyMemberId, insertExternalInfo } from '../../../dal/exportRepo';
import { ExceptionReportIgnoreReasons } from '../../../helpers/exceptionReportRules';
import { markLastExceptionReportAsIgnored } from '../../../dal/exceptionReportRepo';
import { markMemberAsRemoved, removePartyAdditionalInfo } from '../../../dal/partyRepo';
import { archiveExternalInfo } from '../../externalPartyMemberInfo';
import { logPartyMemberRemoved } from '../../helpers/workflows';
import { getContactsInfoByEmail } from '../../../dal/contactInfoRepo';
import { isEmailValid } from '../../../../common/helpers/validations';
import { formatPhoneNumberForDb } from '../../../helpers/phoneUtils';
import { enhance } from '../../../../common/helpers/contactInfoUtils';

const logger = loggerModule.child({ subType: 'importActiveLeases' });
export const baseRentChargeCode = 'RNT';

export const shouldOverrideContactInfo = config.import.overrideContactInfo;

export const cleanAndFormatName = fullName => {
  const formattedfullName = fullName
    // remove anything between parentheses and nested parentheses
    .replace(/\s+\(.*\)\s+/g, ' ')
    // remove everything after (in case parenthesis is not closed)
    .replace(/\(.*$/g, ' ')
    // remove other symbols and numbers
    .replace(/[|#^!?%*_={}:"&;$%@"-<>.~+,0-9]/g, ' ')
    // multiple spaces into one
    .replace(/\s+/g, ' ')
    .trim()
    .split(' ')
    .join('%');

  return `%${formattedfullName}%`;
};
export const cleanAndFormatRevaName = fullname => cleanAndFormatName(fullname).replace(/%/g, ' ');

export const getFullName = receivedResident =>
  removeSpaces([receivedResident.firstName, receivedResident.middleInitial, receivedResident.lastName].filter(x => x).join(' '));

export const getFormattedNamesForQuery = receivedResident => {
  const nameWithMiddleInitial = cleanAndFormatName(
    [receivedResident.firstName, receivedResident.middleInitial, receivedResident.lastName].filter(x => x).join(' '),
  );
  const nameWithoutMiddleInitial = cleanAndFormatName([receivedResident.firstName, receivedResident.lastName].filter(x => x).join(' '));
  return { nameWithMiddleInitial, nameWithoutMiddleInitial };
};

export const getRecurringCharges = charges => charges.filter(c => c.code !== baseRentChargeCode && parseInt(c.amount, 10) > 0);
export const getConcessions = charges => charges.filter(c => c.code !== baseRentChargeCode && parseInt(c.amount, 10) < 0);

export const parseDate = (date, timezone) => parseAsInTimezone(date, { format: YEAR_MONTH_DAY_FORMAT, timezone }).toISOString();

export const getFormattedCharges = (charges, timezone) =>
  charges.map(({ code, description, amount, startDate, quantity, endDate }) => ({
    code,
    description: '',
    displayName: description,
    amount: amount > 0 ? amount : amount * -1,
    startDate: parseDate(startDate, timezone),
    endDate: endDate ? parseDate(endDate, timezone) : null,
    quantity: quantity || 1,
  }));

export const getContactInfosValueByType = contactInfo => {
  const emails = contactInfo
    .filter(ci => ci.type === DALTypes.ContactInfoType.EMAIL)
    .map(email => email.value)
    .sort();
  const phones = contactInfo
    .filter(ci => ci.type === DALTypes.ContactInfoType.PHONE)
    .map(phone => phone.value)
    .sort();
  return { emails, phones };
};

export const getDeletedMembersFromImport = (revaResidents, receivedResidents) => {
  const deletedResidentsIds = [];
  const deletedChildrenIds = [];

  revaResidents.forEach(revaResident => {
    const receivedResident = receivedResidents.find(e => e.id === revaResident.externalId || e.id === revaResident.externalRoommateId);
    if (!receivedResident) {
      if (revaResident.childId) {
        deletedChildrenIds.push(revaResident.childId);
        return;
      }
      deletedResidentsIds.push(revaResident.partyMemberId);
    }
  });
  return { deletedResidentsIds, deletedChildrenIds };
};

const getInventoryExternalId = (propertyExternalId, buildingId, unitId) => (buildingId ? [propertyExternalId, buildingId, unitId].join('-') : unitId);

export const getInventoryByExternalId = (inventories, propertyExternalId, buildingId, unitId) => {
  const receivedInventoryExtId = getInventoryExternalId(propertyExternalId, buildingId, unitId);
  return (
    inventories.find(i =>
      buildingId
        ? getInventoryExternalId(i.propertyExternalId, i.buildingExternalId, i.inventoryExternalId) === receivedInventoryExtId
        : i.inventoryExternalId === receivedInventoryExtId,
    ) || {}
  );
};

export const isLeaseMoveInComplete = entry => {
  const { status } = entry.rawData;
  return status === DALTypes.PartyStateType.RESIDENT;
};

export const isRenewalLetterSent = data => {
  const { wasExternalRenewalLetterSent } = data;
  return wasExternalRenewalLetterSent === 'Y';
};

export const isPartyMovingOut = entry => entry.rawData.leaseVacateDate || entry.rawData.isUnderEviction;

export const isLeaseMoveOutComplete = entry => {
  const { status } = entry.rawData;
  return status === DALTypes.PartyStateType.PASTRESIDENT;
};

export const getPartyWorkflow = (partyWorkflows, workflowName, forceSync = false) => {
  const filteredWorkflows = forceSync
    ? partyWorkflows.filter(item => item.workflowName === workflowName)
    : partyWorkflows.filter(item => item.workflowName === workflowName && item.workflowState === DALTypes.WorkflowState.ACTIVE);

  const enhancedSortedWorkflows = filteredWorkflows
    .map(item => ({
      ...item,
      createdAt: item.seedPartyCreatedAt || item.createdAt,
    }))
    .sort((a, b) => -toMoment(a.createdAt).diff(toMoment(b.createdAt)));

  const activeWorkflow = (enhancedSortedWorkflows || []).find(item => item.workflowState === DALTypes.WorkflowState.ACTIVE);

  return activeWorkflow || enhancedSortedWorkflows[0];
};

export const isLeaseMoveInNotCompleted = entry => {
  const { status } = entry.rawData;
  return status === DALTypes.PartyStateType.FUTURERESIDENT;
};

export const getEnhancedContactInfoWithSmsInfo = async (ctx, personId, contactInfos) => {
  try {
    return !shouldOverrideContactInfo ? await enhanceContactInfoWithSmsInfo(ctx, contactInfos) : contactInfos;
  } catch (error) {
    logger.error({ ctx, personId }, 'getEnhancedContactInfoWithSmsInfo - error');
    return contactInfos;
  }
};

export const validatePhone = phone => (shouldOverrideContactInfo ? { valid: true } : validatePhoneNumber(phone));

const overrideEmail = email => {
  const emailSubstr = email.substring(0, email.lastIndexOf('.')).replace('@', '_');
  return `cust+${emailSubstr}@reva.tech`;
};

const overridePhone = phone => {
  const last7Digits = trim(phone).slice(phone.length - 7);
  return `1111${last7Digits}`;
};

export const overrideContactInfo = persons =>
  persons?.length &&
  persons.map(person => ({ ...person, email: person?.email ? overrideEmail(person.email) : null, phone: person?.phone ? overridePhone(person.phone) : null }));

export const getExternalRoommateId = (ctx, { externalId, isPrimary }) => (!isPrimary && ctx.backendMode === DALTypes.BackendMode.YARDI ? externalId : null);

export const isPartyTransferred = (receivedInventoryId, currentInventoryId) => receivedInventoryId !== currentInventoryId;

export const archivePartiesOnTransfer = async (ctx, activeLeasePartyId, partyWorkflows) => {
  await archiveParty(ctx, {
    partyId: activeLeasePartyId,
    workflowName: DALTypes.WorkflowName.ACTIVE_LEASE,
    archiveReasonId: DALTypes.ArchivePartyReasons.RESIDENTS_TRANSFERRED,
  });

  const newLeaseWorkflow = getPartyWorkflow(partyWorkflows, DALTypes.WorkflowName.NEW_LEASE);
  const newLeasePartyId = newLeaseWorkflow?.id;
  !!newLeasePartyId &&
    (await archiveParty(ctx, {
      partyId: newLeasePartyId,
      workflowName: DALTypes.WorkflowName.NEW_LEASE,
      archiveReasonId: DALTypes.ArchivePartyReasons.RESIDENTS_TRANSFERRED,
    }));
};

export const getActiveLeaseData = async (ctx, partyWorkflows) => {
  const activeLeaseWorkflow = getPartyWorkflow(partyWorkflows, DALTypes.WorkflowName.ACTIVE_LEASE);
  const activeLeaseData = activeLeaseWorkflow && (await getActiveLeaseWorkflowDataByPartyId(ctx, activeLeaseWorkflow.id));
  return activeLeaseData && { ...activeLeaseData, partyGroupId: activeLeaseWorkflow.partyGroupId, leaseType: activeLeaseWorkflow.leaseType };
};

export const handleUpdatedExternalInfo = async (ctx, externalToArchive, newExternalInfo) => {
  const { partyId, memberId, propertyId } = externalToArchive;
  const { externalId = null, externalProspectId = null, externalRoommateId = null, isPrimary = false, isChild = false, metadata = {} } = newExternalInfo;
  logger.trace({ ctx, partyId, memberId, externalId, externalProspectId, externalRoommateId, isPrimary, isChild }, 'handleUpdatedExternalInfo');

  await archiveExternalInfoByPartyMemberId(ctx, memberId);

  await insertExternalInfo(ctx, {
    partyId,
    partyMemberId: isChild ? null : memberId,
    childId: isChild ? memberId : null,
    externalId,
    externalProspectId,
    externalRoommateId,
    isPrimary,
    metadata,
    propertyId,
  });
};

export const getPhoneNumberAndExtension = phone => {
  if (!phone) return {};

  let phoneNumber = phone;
  let phoneNumberExtension;

  if (phone.includes('x')) {
    [phoneNumber, phoneNumberExtension] = phone.split('x');
  }
  return { phoneNumber: getOnlyDigitsFromPhoneNumber(phoneNumber), phoneNumberExtension };
};

export const insertExternalInfoAndIgnorePreviousER = async (
  ctx,
  { sameMemberFoundInParty, receivedResident, renewalPartyId, isPrimary, propertyId, timezone },
) => {
  logger.trace(
    { ctx, sameMemberFoundInParty, receivedResidentId: receivedResident.id, renewalPartyId, isPrimary },
    'insertExternalInfoAndIgnorePreviousExceptionReport',
  );
  const externalRoommateId = getExternalRoommateId(ctx, { externalId: receivedResident.id, isPrimary });

  await insertExternalInfo(ctx, {
    partyId: renewalPartyId,
    partyMemberId: sameMemberFoundInParty.matchedPartyMemberId,
    childId: sameMemberFoundInParty.childId,
    externalId: externalRoommateId ? null : receivedResident.id,
    externalProspectId: null,
    externalRoommateId,
    isPrimary,
    metadata: {},
    propertyId,
  });

  const ignoreReason = {
    date: now({ timezone }).format(YEAR_MONTH_DAY_FORMAT),
    reason: ExceptionReportIgnoreReasons.PERSON_ALREADY_ADDED_IN_RENEWAL.description,
  };

  await markLastExceptionReportAsIgnored(ctx, {
    externalId: receivedResident.id,
    ignoreReason,
    ruleId: ExceptionReportIgnoreReasons.PERSON_ALREADY_ADDED_IN_RENEWAL.ruleId,
  });
};

export const wasEntryNotUpdated = (newEntry, entry) => isEqual(newEntry.rawData, entry.rawData);

export const removeResident = async (ctx, activeRevaMembersInfo, deletedResidentId) => {
  const externalInfo = activeRevaMembersInfo.find(info => info.partyMemberId === deletedResidentId);
  const removedMember = await markMemberAsRemoved(ctx, deletedResidentId);
  await archiveExternalInfo(ctx, externalInfo);
  await logPartyMemberRemoved(ctx, removedMember);
};

export const removeChild = async (ctx, activeRevaMembersInfo, deletedChildId) => {
  const externalInfo = activeRevaMembersInfo.find(info => info.childId === deletedChildId);
  await removePartyAdditionalInfo(ctx, deletedChildId);
  await archiveExternalInfo(ctx, externalInfo);
};

export const buildContactInfo = async (ctx, receivedResident) => {
  const { id, email, phone } = receivedResident;
  const contactInfos = [];

  if (email) {
    const [existing] = await getContactsInfoByEmail(ctx, email);
    const isValid = isEmailValid(email);

    if (isValid && !existing) {
      contactInfos.push({
        type: DALTypes.ContactInfoType.EMAIL,
        value: email,
        imported: true,
      });
    } else {
      !isValid
        ? logger.error({ ctx, id, email }, 'buildContactInfo - invalid email')
        : logger.error({ ctx, id, email }, 'buildContactInfo - email already exists');
    }
  }

  if (phone) {
    const { phoneNumber, phoneNumberExtension } = getPhoneNumberAndExtension(phone);
    const { valid, reason } = validatePhone(phoneNumber);
    if (valid) {
      contactInfos.push({
        type: DALTypes.ContactInfoType.PHONE,
        value: formatPhoneNumberForDb(phoneNumber),
        imported: true,
        metadata: phoneNumberExtension ? { phoneNumberExtension } : {},
      });
    } else {
      logger.error({ ctx, id, phone, reason }, 'buildContactInfo - invalid phone');
    }
  }

  return enhance(contactInfos);
};
