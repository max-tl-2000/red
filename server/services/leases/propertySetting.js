/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { contains, LeaseDocuments } from '../../../common/helpers/leaseDocuments';
import { formatUnitAddress } from '../../../common/helpers/addressUtils';
const COVE = 'cove';
const LARK = 'lark';
const WOOD = 'wood';
const SWPARKME = 'swparkme';
const SHARON = 'sharon';
const SHORE = 'shore';

const alwaysIncluded = matches => name => matches.find(m => contains(name, m));

// TODO: read this from property settings after import
const propertyLeaseSettings = {
  [COVE]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Sat',
      officeDays2: 'Sun',
      officeDays3: '',
      officeHours1: '9am - 6pm',
      officeHours2: '11am - 5pm',
      officeHours3: '',
    },
    leadDisclosureClause: 'DO NOT',
    asbestosLocations: 'Drywall & Stucco',
    rentDueDay: 'fifth (5th)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy: 'Barbecue grills are permitted with 1 propane tank weighing no more than 1 pound.',
    alwaysIncluded: alwaysIncluded([]),
    propertyPhone: '',
    unitAddress: formatUnitAddress,
  },
  [LARK]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Sat',
      officeDays2: 'Sun',
      officeDays3: '',
      officeHours1: '10am - 6pm',
      officeHours2: '10am - 5pm',
      officeHours3: '',
    },
    leadDisclosureClause: 'DO NOT',
    asbestosLocations: 'Vinyl flooring, sloped roofing material, roofing mastic and fire doors',
    rentDueDay: 'fifth (5th)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy:
      'The use of any charcoal burner, liquid petroleum gas fueled or any other open flame cooking device are prohibited in tenant’s unit or on their patio/balcony.',
    alwaysIncluded: alwaysIncluded([]),
    propertyPhone: '',
    unitAddress: formatUnitAddress,
  },
  [WOOD]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Fri',
      officeDays2: 'Sat',
      officeDays3: '',
      officeHours1: '9am - 6pm',
      officeHours2: '10am - 5pm',
      officeHours3: '',
    },
    leadDisclosureClause: 'DO NOT',
    asbestosLocations: 'Walls, Ceilings, Floor Tiles, Wall/Floor Insulation',
    rentDueDay: 'fifth (5th)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy:
      'The use of any charcoal burner, liquid petroleum gas fueled or any other open flame cooking device are prohibited in tenant’s unit or on their patio/balcony.',
    alwaysIncluded: alwaysIncluded([]),
    propertyPhone: '',
    unitAddress: formatUnitAddress,
  },
  [SWPARKME]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Fri',
      officeDays2: 'Sat',
      officeDays3: '',
      officeHours1: '9am - 6pm',
      officeHours2: '',
      officeHours3: '',
    },
    leadDisclosureClause: '',
    asbestosLocations:
      'in the tile floors on the second through top floors, halls and elevators of all the high-rise apartment buildings; insulation on the piping and heaters in the boiler rooms, laundry rooms and in the basement hallways; insulation for the heating system duct work in garden apartments in the heater closets; and in some apartment kitchen floor coverings',
    rentDueDay: 'first (1st)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy: '',
    alwaysIncluded: alwaysIncluded([LeaseDocuments.PET_AGREEMENT_ADDENDUM]),
    propertyPhone: '(415) 405-4690',
    unitAddress: formatUnitAddress,
  },
  [SHARON]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Sat',
      officeDays2: 'Sun',
      officeDays3: '',
      officeHours1: '9am - 6pm',
      officeHours2: '10am - 5pm',
      officeHours3: '',
    },
    leadDisclosureClause: 'DO',
    asbestosLocations: 'drywall, stucco, and select floor adhesive',
    rentDueDay: 'first (1st)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy:
      'The use of any charcoal burner, liquid petroleum gas fueled or any other open flame cooking device are prohibited in tenant’s unit or on their patio/balcony.',
    alwaysIncluded: alwaysIncluded([]),
    propertyPhone: '',
    unitAddress: formatUnitAddress,
  },
  [SHORE]: {
    officeWorkingTime: {
      officeDays1: 'Mon - Sat',
      officeDays2: 'Sun',
      officeDays3: '',
      officeHours1: '9am - 6pm',
      officeHours2: '10am - 5pm',
      officeHours3: '',
    },
    leadDisclosureClause: 'DO',
    asbestosLocations: '',
    rentDueDay: 'first (1st)',
    countersignerDescriptor: "Owner or Owner's Representative",
    barbecuePolicy:
      'The use of any charcoal burner, liquid petroleum gas fueled or any other open flame cooking device are prohibited in tenant’s unit or on their patio/balcony.',
    alwaysIncluded: alwaysIncluded([]),
    propertyPhone: '',
    unitAddress: formatUnitAddress,
  },
};

// TODO: make this editable via UI
const defaultPropertySettings = {
  officeWorkingTime: {
    officeDays1: 'Mon - Fri',
    officeDays2: 'Sat',
    officeDays3: '',
    officeHours1: '9am - 6pm',
    officeHours2: '10am - 5pm',
    officeHours3: '',
  },
  leadDisclosureClause: 'DO NOT',
  asbestosLocations: '',
  rentDueDay: 'fifth (5th)',
  countersignerDescriptor: "Owner or Owner's Representative",
  barbecuePolicy:
    'The use of any charcoal burner, liquid petroleum gas fueled or any other open flame cooking device are prohibited in tenant’s unit or on their patio/balcony.',
  alwaysIncluded: alwaysIncluded([]),
  propertyPhone: '',
  unitAddress: formatUnitAddress,
};

export const naturalGasProvidedProperties = ['11500', '12982'];
export const submeterAddendumProperties = ['13780', '11500', '11501', '12982'];
export const utilityAddendumProperties = ['11190', '14160'];

export const getLeaseSettingsForProperty = propertyName => propertyLeaseSettings[propertyName] || defaultPropertySettings;
