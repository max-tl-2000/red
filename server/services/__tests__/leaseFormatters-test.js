/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { formatPropertyAddress } from '../../../common/helpers/addressUtils';
import { getLeaseSettingsForProperty } from '../leases/propertySetting';

const inventoryAddress = '370 Channing Way';
const generateTestData = (propertyName, hasInventoryAddress = false) => ({
  type: 'Unit',
  name: '164',
  address: hasInventoryAddress ? inventoryAddress : '',
  building: {
    address: {
      addressLine1: '2300 Lincoln village circle',
      city: 'Tiburon',
      state: 'CA',
      postalCode: '94920',
    },
  },
});

describe('services/leaseFormatters', () => {
  describe('When formatting a unit address using formatUnitAddress', () => {
    it('should return empty if no building prop found', () => {
      const propertySettings = getLeaseSettingsForProperty();
      const unitAddress = propertySettings.unitAddress({});
      expect(unitAddress).toEqual('');
    });

    it('should return empty if undefined is provided', () => {
      const propertySettings = getLeaseSettingsForProperty();
      const unitAddress = propertySettings.unitAddress(undefined);
      expect(unitAddress).toEqual('');
    });

    it('should use the inventory address when the inventory has one set', () => {
      const coveInvetory = generateTestData('cove', true);
      const propertySettings = getLeaseSettingsForProperty('cove');
      const unitAddress = propertySettings.unitAddress(coveInvetory);
      expect(unitAddress).toMatchSnapshot();
    });

    it('should use the building address when the inventory address is not set', () => {
      const larkInvetory = generateTestData('lark', false);
      const propertySettings = getLeaseSettingsForProperty('lark');
      const unitAddress = propertySettings.unitAddress(larkInvetory);
      expect(unitAddress).toMatchSnapshot();
    });
  });

  describe('when formatting an address using formatPropertyAddress', () => {
    it('should construct an address string from the provided data omitting blanks', () => {
      const address = formatPropertyAddress({
        addressLine1: '373 E El Camino Real',
        city: 'Sunnyvale',
        state: 'CA',
        postalCode: '95041',
      });
      expect(address).toMatchSnapshot();

      const address2 = formatPropertyAddress({
        addressLine1: '373 E El Camino Real',
        addressLine2: 'Apt 1207',
        city: 'Sunnyvale',
        state: 'CA',
        postalCode: '95041',
      });
      expect(address2).toMatchSnapshot();
    });
  });
});
