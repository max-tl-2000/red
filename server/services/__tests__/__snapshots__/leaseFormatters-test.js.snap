// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services/leaseFormatters When formatting a unit address using formatUnitAddress should use the building address when the inventory address is not set 1`] = `"2300 Lincoln village circle, Unit 164, Tiburon, CA 94920"`;

exports[`services/leaseFormatters When formatting a unit address using formatUnitAddress should use the inventory address when the inventory has one set 1`] = `"370 Channing Way, Tiburon, CA 94920"`;

exports[`services/leaseFormatters when formatting an address using formatPropertyAddress should construct an address string from the provided data omitting blanks 1`] = `"373 E El Camino Real, Sunnyvale, CA 95041"`;

exports[`services/leaseFormatters when formatting an address using formatPropertyAddress should construct an address string from the provided data omitting blanks 2`] = `"373 E El Camino Real, Apt 1207, Sunnyvale, CA 95041"`;
